### PR TITLE
[FLINK-4611] [kinesis] Make "AUTO" credential provider as default for Kinesis Connector

### PR DESCRIPTION
**Pull request by tony810430**

### CODE REVIEW

**Comment by StephanEwen** [comment]
Looks like a good change to me.
Would be great if someone could try that out on an EMR/Kinesis cluster and validate it practically.

**Comment by tzulitai** [comment]
I've tested the `AUTO` config behaviour, and it works as expected in EMR / EC2 environments, so the Kinesis API part should be working fine.

Let me summarize the new behaviour after the proposed changes (@tony810430 please correct me if I've got anything wrong):
- If no credential provider type is provided, `AUTO` is used.
- If the user has provided `AWS_SECRET_ACCESS_KEY` and `AWS_ACCESS_KEY_ID`, regardless of the credential provider type provided (i.e. even if it's set to `AUTO` / `SYS_PROPS` etc.), the `BASIC` type will be used.

The second behaviour seems a little odd to me; in general, I think the `AWS_CREDENTIALS_PROVIDER` setting should have higher precedence.

Also, logic between `validateAwsConfiguration` and `AWSUtils.getCredentialsProvider()` seems a bit too "coupled" now with the proposed changes. I think we should probably fallback the value of `AWS_CREDENTIALS_PROVIDER` for the user in `validateAwsConfiguration`. What I have in mind for `validateAwsConfiguration`:

- if the user hasn't set any value for `AWS_CREDENTIALS_PROVIDER`, we set it for them - if access key and secret access key exists, set `BASIC`, otherwise just use `AUTO` (the default).
- if the user has a value for `AWS_CREDENTIALS_PROVIDER`, just leave it as is.

In `AWS_CREDENTIALS_PROVIDER`, we can then just start of by reading the `AWS_CREDENTIALS_PROVIDER` property and instantiate credentials accordingly.

@tony810430 what do you think?

**Comment by tony810430** [comment]
The new behavior is as what you have said. The second behavior is for letting configuration in previous version work like before, otherwise job will fail in this version if it couldn't get the aws credential from local.

The reason why I implement the logic in `validateAwsConfiguration` and `AWSUtils.getCredentialsProvider()` like this is because I thought config properties should be immutable and validation shouldn't do anything more than checking if theses properties were correct. Therefore, it is reasonable for me to remain the current implementation, but I probably need to make the parameter in `validateAwsConfiguration`  be immutable.

**Comment by tzulitai** [comment]
Thanks for the clarification.

> 
If the user has provided AWS_SECRET_ACCESS_KEY and AWS_ACCESS_KEY_ID, regardless of the credential provider type provided (i.e. even if it's set to AUTO / SYS_PROPS etc.), the BASIC type will be used.

I misunderstood your code before, so scratch this, sorry about that.
Your recent updates look good, +1 to merge after Travis finishes a test run.

File: flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumerTest.java [hunk 1]
```diff
@@ -87,18 +87,6 @@ public void testUnrecognizableAwsRegionInConfig() {
 	}
 
 	@Test
-	public void testCredentialProviderTypeDefaultToBasicButNoCredentialsSetInConfig() {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumerTest.java]: We should actually add a test to replace this, for the new behaviours. @tony810430 would you like to add it?

Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumerTest.java]: Makes sense, just realized that we don't really have unit tests for `AWSUtil` right now, and the removed test simply corresponds to the reduced behaviour of `validateAwsConfiguration()`. Thanks for pointing this out!

**Comment by tzulitai** [comment]
Merging ...

**Pull request closed without merging**